### PR TITLE
Introduce Storefront Offers

### PR DESCRIPTION
**Pull request by justnero**

### CODE REVIEW

File: spec/components/requestBodies/Offer.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      type: object
+      required:
+        - currency
+        - websiteId
+        - charges
+      properties:
+        currency:
+          allOf:
+            - $ref: "#/components/schemas/CurrencyCode"
+        websiteId:
```
Comment (adamaltman) [inline spec/components/requestBodies/Offer.yaml]: Why is this required and not just a possible restriction?

File: spec/components/schemas/Offers/Restrictions/total-purchases.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+allOf:
+  - $ref: "#/components/schemas/OfferRestriction"
+  - type: object
+    required:
+      - amount
+    properties:
+      amount:
+        description: The number of purchases allowed on the offer
```
Comment (adamaltman) [inline spec/components/schemas/Offers/Restrictions/total-purchases.yaml]: This whole restriction part doesn't make sense to me.

File: spec/components/schemas/Offers/OfferRestriction.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+discriminator:
+  propertyName: type
+required:
+  - type
+properties:
+  type:
+    type: string
+    enum:
+      - total-purchases
+      - gateway-accounts
```
Comment (adamaltman) [inline spec/components/schemas/Offers/OfferRestriction.yaml]: This doesn't make sense.  I need to re-tell the use case.

File: spec/paths/offers.yaml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+              $ref: "#/components/schemas/Offer"
+    401:
+      $ref: "#/components/responses/AccessForbidden"
+post:
+  tags:
+    - Offers
+  summary: Create an Offer
+  operationId: PostOffer
+  description: |
+    Create an Offer
```
Comment (adamaltman) [inline spec/paths/offers.yaml]: This is where you would write in human-readable sentences and explain why someone would want to create an offer. And you should add the tag to the tags section in openapi.yaml with a description including the what/why explained there.

File: spec/components/requestBodies/Offer.yaml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+          description: The website's ID
+          allOf:
+            - $ref: "#/components/schemas/ResourceId"
+        charges:
+          description: The charges applied customer
```
Comment (adamaltman) [inline spec/components/requestBodies/Offer.yaml]: This seems an incorrect description.  As I understand the customer may select one of these.

File: spec/components/schemas/Offers/OfferCharge.yaml [hunk 6]
```diff
@@ -0,0 +1,33 @@
+type: object
+required:
+  - name
+  - pricing
+  - quantity
+properties:
+  name:
+    type: string
+    description: The name of the charge (used in integrations only)
```
Comment (adamaltman) [inline spec/components/schemas/Offers/OfferCharge.yaml]: Does it need to be unique? How is it used in integrations?

File: spec/components/schemas/Offers/OfferRestriction.yaml [hunk 7]
```diff
@@ -0,0 +1,11 @@
+type: object
+discriminator:
```
Comment (adamaltman) [inline spec/components/schemas/Offers/OfferRestriction.yaml]: When possible, avoid the discriminator. I will request every API that has it should use `oneOf` instead.

File: spec/paths/offers.yaml (hunk 4 above)
Comment (adamaltman) [reply spec/paths/offers.yaml]: I forgot to mention, and when you want to do it.

File: spec/components/requestBodies/Offer.yaml (hunk 5 above)
Comment (justnero) [reply spec/components/requestBodies/Offer.yaml]: No, all charges are applied at the same time. The only thing customer can choose is quantity

Comment (adamaltman) [reply spec/components/requestBodies/Offer.yaml]: I don't understand how it works.  That means this documentation is incomplete.

File: spec/components/requestBodies/Offer.yaml (hunk 1 above)
Comment (justnero) [reply spec/components/requestBodies/Offer.yaml]: It doesn't make sense for `websiteId` to be optional here. When it is required we can use it for transaction creation, but as an optional value it doesn't worth. For limiting Offer visibility on different websites we have ACL

Comment (adamaltman) [reply spec/components/requestBodies/Offer.yaml]: I think you misunderstand me here.  I'm not talking about making this optional.  I'm talking about removing it completely -- and then adding a possible restriction type for website.  As you propose, this solution will not work for certain clients.  I'm happy to discuss it on slack.

File: spec/paths/offers@{id}.yaml [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+      $ref: "#/components/responses/NotFound"
+    422:
+      $ref: "#/components/responses/InvalidDataError"
+delete:
+  tags:
+    - Offers
+  summary: Delete an Offer
+  operationId: DeleteOffer
+  description: |
+    Delete an Offer with predefined identifier string
```
Comment (ashkarpetin) [inline spec/paths/offers@{id}.yaml]: Does delete remove the record from table completely or it just changes status to deactivated?

**Comment by justnero** [comment]
Offers are not relevant for our merchants in current version
